### PR TITLE
Transformers/DETR integration

### PR DESCRIPTION
**Pull request by jveitchmichaelis**
This PR includes rudimentary support for transformers object detection models, specifically those supported by `AutoModelForObjectDetection`. I've trained a simple DeformableDetr model as a demonstration using the latest public release of MillionTrees which is hosted here (for testing): https://huggingface.co/joshvm/milliontrees-detr

The PR fixes a few issues in the process:

- Fixed an issue with albumentations versioning which can be rebased once the other PR is approved
- Adds dependencies for transformers and timm. I think as I used `uv` to add the pyproject line, it should be compatible with the python versions that we support.
- The `config.score_thresh` parameter is moved to be a top-level config item as there's no need for it to be specific to retinanet, and other places it's referenced have been updated. This is also used for post-processing results from DETR.
- Removed a comment in the config file that incorrectly suggests that threshold is NMS-related
- Removed some docstrings in models that were out of date (referring to constructor arguments that no longer exist)
- Adds a model class currently called `DeformableDETR`, but really it could support a variety of models. These can be separate files, but we would likely want to re-use the wrapper.
- A small improvement to plot_results to allow an image path to provided instead of an array instead of erroring.

An advantage of the current approach is that it requires no changes to `main` at the moment, the implementation is isolated to `models/detr.py`. I need to check compatibility with other prediction routes (like via the trainer) but the `forward` function returns [boxes, labels, scores] in a DeepForest-friendly format without any modification. The wrapper class is a bit awkward, but necessary because of the way that `Model` and sub-classes are currently implemented (since we need to run pre- and post-process which is transformers specific).

The model and processor (and their config as it relates to DeepForest) should be coupled, so I don't think we need to worry about exposing the details. So I think it's OK to have both loaded with `config.model.name` and keep the config fixed, rather than allow users to mess with image resizing. Other options like rescaling (1/255) and normalization should be preset so that the processor works with our dataset + loader.

NMS for `predict_image` doesn't seem to be working though, and I'm not sure why (@bw4sz?), see the following test script:

```python
from deepforest.main import deepforest
from deepforest.visualize import plot_results

m = deepforest(config_args={"model": {"name": "joshvm/milliontrees-detr"},
                            "score_thresh": 0.25, # This does work, at least.
                            "architecture": "detr"})

image_path = "src/deepforest/data/2018_SJER_3_252000_4107000_image_477.tif"

results = m.predict_image(path=image_path)

plot_results(results, image=image_path)
```

image

Training is going to need a bigger refactor (probably) because the various Lightning `step`s in `main` make assumptions that might be tied to torchvision-type models.

TODO:

- Test suite. And fix the stuff that's currently breaking.
- Tidy up docstrings etc.
- Check prediction with other functions.
- Think about how we can enable training as well, vs just loading models.
- Evaluation check vs retinanet.
- Docs for what is required to pull in a model from HF and configuration parameters.

### CODE REVIEW

**Comment by jveitchmichaelis** [comment]
Changed this to explicitly import DeformableDetr. I've tested basic forward passes with dummy data and data loaded from a BoxDataset. Model loading is probably broken because we `load_model` probably requires the checkpoint to be generated from `DeepForest`'s training pipeline. You can get around this by loading weights with `create_model`, which default to the MS-COCO version (like Faster-RCNN and retinanet).

**Comment by jveitchmichaelis** [comment]
Will revert the NMS config changes, don't think that needs to be here. Same goes for some minor typos etc in the test cases for Faster-RCNN; that can be a separate PR.

File: src/deepforest/model.py [hunk 1]
```diff
@@ -20,17 +20,14 @@ class Model():
...
-    Returns:
-        model: a pytorch nn module
+        config (DictConfig): DeepForest config settings object
     """
 
     def __init__(self, config):
 
         # Check for required properties and formats
         self.config = config
+        self.nms_thresh = None  # Required for some models but not all
```
Comment (henrykironde) [inline src/deepforest/model.py:30]: Do we really need this here? self.nms_thresh = Non

**Review by henrykironde — COMMENTED** [review]
These changes look good for now. Hope we are planning on incorporating parallelization in this design?

File: src/deepforest/main.py [hunk 2]
```diff
@@ -136,16 +137,22 @@ def load_model(self, model_name="weecology/deepforest-tree", revision='main'):
...
-        self.create_model()
         loaded_model = self.from_pretrained(model_name, revision=revision)
         self.label_dict = loaded_model.label_dict
         self.model = loaded_model.model
         self.numeric_to_label_dict = loaded_model.numeric_to_label_dict
 
         # Set bird-specific settings if loading the bird model
         if model_name == "weecology/deepforest-bird":
-            self.config.retinanet.score_thresh = 0.3
+            self.config.score_thresh = 0.3
```
[the remark this answers is not in the document]
Comment (jveitchmichaelis) [reply src/deepforest/main.py:155]: Yeah I think this is an override for the bird model, but it would be more correct to have a different config entirely and not hard code it here.

File: src/deepforest/model.py (hunk 1 above)
Comment (jveitchmichaelis) [reply src/deepforest/model.py:30]: Hmm probably not, let me check

**Comment by jveitchmichaelis** [comment]
On NMS: According to the DETR authors, the algorithm is supposed to be "NMS-Free", but clearly the post-processed result from `transformers` still returns a lot of overlapping predictions (and they're not no-object).

This makes sense for training, because the ground truth/prediction matcher is 1:1 by definition (via Hungarian Matching/linear sum assignment), so you don't need to first run NMS to clean up the boxes before calculating the loss. The model _should_ learn to not over-predict.

But when there's nothing to compare against, all the predictions are retained (unless you set a higher confidence score).

See https://github.com/facebookresearch/detr/issues/72

**Comment by jveitchmichaelis** [comment]
Will begin to revert changes that are moved to other PRs.

**Comment by bw4sz** [comment]
What's the latest here with NMS? predict_image doesn't perform NMS, only when there are multiple classes present.

https://github.com/weecology/DeepForest/blob/c53f0904725fbba1e08d873032a229bf303a6315/src/deepforest/predict.py#L43C1-L44C1

because torchvision does nms internally in retinanet.

File: src/deepforest/models/DeformableDetr.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+                if isinstance(images, list) else [images.shape[-2:]])
+        else:
+            return preds.loss_dict
+
+
+class Model(Model):
+
+    def __init__(self, config, **kwargs):
+        """
+        Args:
```
[the remark this answers is not in the document]
Comment (henrykironde) [reply src/deepforest/models/DeformableDetr.py:96]: Naming a class as its parent class is little confusing `class Model(Model): ` would it be to much to change to something different ?

**Review by henrykironde — COMMENTED** [review]
I've added a few comments. For the tests, I think we should include proper assertions to validate behavior more clearly. Also, we should follow Python naming conventions and rename the file from `DeformableDetr.py` to `deformable_detr.py`

File: tests/test_detr.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
+# test Transformers/detr
+from deepforest.models import DeformableDetr
+from deepforest import utilities
+from deepforest.datasets.training import BoxDataset
+from deepforest import get_data
+import pytest
+import torch
+from PIL import Image
+import numpy as np
```
[the remark this answers is not in the document]
Comment (jveitchmichaelis) [reply tests/test_detr.py:9]: Do we have guidelines for this? I normally use `isort` in pre-commit.

File: tests/test_detr.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+import os
+
+@pytest.fixture()
+def config():
+    config = utilities.load_config()
+    config.model.name = "joshvm/milliontrees-detr"
```
[the remark this answers is not in the document]
Comment (jveitchmichaelis) [reply tests/test_detr.py:15]: Sure, let me see if this is the best checkpoint I have. I think there's likely a better one that doesn't over-predict.

**Comment by jveitchmichaelis** [comment]
@henrykironde - review summary:

- Will add a boolean return value to `check_model` - that function also asserts internally, but it doesn't return anything that we could use for tests.
- For the other test we can use predictions with random weights, they'll just have very low scores (fine for the test case)
- I'm proposing a refactor of weight loading to better support new models: https://github.com/weecology/DeepForest/pull/1083
- If we go with snake case, we should also think about moving to `faster_rcnn` and making similar changes. That may be breaking?
- I've run an import sort on the test file - for consistency we ought to also do this for other modules. Alternatively consider automatically doing this on commit. Yapf explicitly doesn't support this, but isort can be run in combination with many formatters in a deterministic way.
- Removed unused imports from test_retinanet.

File: tests/test_detr.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (henrykironde) [reply tests/test_detr.py:9]:
```Python
import os

import cv2
import numpy as np
import rasterio
import torch
import torch.nn.functional as F
import torchmetrics
from pytorch_lightning import LightningModule, Trainer
from torchvision import models, transforms
from torchvision.datasets import ImageFolder
```

I usually use Pycharm, it organizes imports or optimizes imports.  But we should have 

Organized by type:
```
Standard library imports first (os)
Third-party imports second (all the others)
Alphabetical ordering within each groups 
```

**Comment by martibosch** [comment]
Hello @jveitchmichaelis 

> * I've run an import sort on the test file - for consistency we ought to also do this for other modules. Alternatively consider automatically doing this on commit. Yapf explicitly doesn't support this, but isort can be run in combination with many formatters in a deterministic way.
> * Removed unused imports from test_retinanet.

I have also encountered similar issues in #901 and I think it would indeed be a good idea to enforce a consistent styling. I think that there is a big consensus that using ruff is the way to go, so I have drafted #1084.

Happy to discuss this further. Best,
Martí

**Comment by jveitchmichaelis** [comment]
We need to fix the other PR first (weight reloading) as that directly affects this one, but once that's figured out we can merge this.

**Comment by henrykironde** [comment]
I didn't see that comment coming in, but If we have to revert let me know @jveitchmichaelis

**Comment by jveitchmichaelis** [comment]
No worries, it'd be good just to tidy up some of the changes in this commit so it's scoped to just the transformers parts. Standby - I'll take a look tomorrow.

**Comment by jveitchmichaelis** [comment]
You could also try the incantation we use in the Action: `
uv sync --all-extras --dev`
